dask/array/gufunc.py

```diff
@@ -0,0 +1,396 @@
+
+
+def _parse_arg(sig):
+    if not sig:
+        return None
+    assert len(re.findall("^\([_a-zA-Z0-9,]*\)$", sig)) == 1
+    sig = sig[1:-1]
+    dims = sig.split(',')
+    ret = [_parse_dim(d) for d in dims]
+    return tuple(r for r in ret if r is not None)
+
+
+def _parse_args(sig):
+    if not sig:
+        return None
+    assert len(re.findall("^(\([_a-zA-Z0-9,]*\)),?(,?\([_a-zA-Z0-9,]*\))*$", sig)) == 1
+    sigs = re.findall("\([_a-zA-Z0-9,]*\)*", sig)
+    ret = [_parse_arg(s) for s in sigs]
+    return list(r for r in ret if r is not None)
+
+
+def parse_signature(signature):
+    """"
+    Parse gufunc signature into nested tuples/lists structures
+    of the dimension names.
+
+    Parameters
+    ----------
+    signature: str
+        According to the specification of numpy.gufunc signature [2]_
+
+    Returns
+    -------
+    Nested tuples/lists structures of the dimension names.
+
+    Examples
+    --------
+    >>> parse_signature('(i),(i,j)->(i,j)')
+    ([('i',), ('i', 'j')], ('i', 'j'))
+
+    >>> parse_signature('(i,j)->(i),(j)')
+    ([('i', 'j')], [('i',), ('j',)])
+
+    >>> parse_signature('->(i),')
+    ([], [('i',)])
+    """
+    signature = signature.replace(' ', '')
+    assert "->" in signature
+    _in, _out = signature.split("->")
+    _ins = _parse_args(_in)
+    if _ins is None:
+        _ins = []
+    _ins = [i for i in _ins if i is not None]
+    _outs = _parse_args(_out)
+    if _outs and (len(_outs) == 1) and ('),' not in _out):
+        _outs = _outs[0]
+    return _ins, _outs
+
+
+def compile_signature(ins, outs):
+    """"
+    Compile gufunc signature from nested tuples/lists structures
+    of the dimension names.
+
+    Parameters
+    ----------
+    ins: list of tuples of dimension names
+        Each input is a tuple of dimension names
+
+    outs: list of tuples of dimension names or tuple of dimension names
+        For one output tuple of dimension names, for many outputs list of
+        tuple of dimension names
+
+    Returns
+    -------
+    Signature according to the specification of numpy.gufunc signature [2]_
+
+    Examples
+    --------
+    >>> compile_signature([('i',), ('i', 'j')], ('i', 'j'))
+    '(i),(i,j)->(i,j)'
+
+    >>> compile_signature([('i', 'j')], [('i',), ('j',)])
+    '(i,j)->(i),(j)'
+
+    >>> compile_signature([], [('i',),])
+    '->(i),'
+    """
+    for e in concat(ins):
+        if len(_valid_name_re.findall(e)) != 1:
+            raise ValueError('`"{0}"` is not a valid name.'.format(e))
+    ins = ",".join(("({0})".format(",".join(dims)) for dims in ins))
+
+    if outs is None:
+        outs = ""
+    elif isinstance(outs, list):
+        for e in concat(outs):
+            if len(_valid_name_re.findall(e)) != 1:
+                raise ValueError('`"{0}"` is not a valid name.'.format(e))
+        _flag = len(outs) == 1
+        outs = ",".join(("({0})".format(",".join(dims)) for dims in outs))
+        if _flag:
+            outs += ","
+    else:
```

Comment:
I think this clause could be dropped if you only support the versions always takes multiple outputs.